intel/intel-iot-refkit
# meta-refkit-core: switch to C-based refkit-ostree for the initramfs ostree module, and ostree updater.

This patch set switches refkit-ostree, the initial ostree wrapper/helper and updater script, to a C-based implementation. The refkit ostree initramfs module has been updated to use this for parsing the boot loader configuration files and pick the latest ostree deployment to boot into.

The patch set also enables automatic ostree-based image updates, updating the UEFI combo app if necessary, and activating the changes via a reboot. On the first boot after an update, a series of checks are performed to decide if the update is considered successful and working. Additional checks can be easily hooked to the existing set (for instance by services or applications, if necessary). The current set of checks do nothing, just marking the update successful.

Note that if the updater service (refkit-update.service) is enabled (which it is by default), entities that need to prevent (= delay) the shutdown to a more convenient time in the future should use systemd-inhibit, or the corresponding systemd(-logind) interfaces, for doing so. For instance, if you have an interactive shell (or a login session via ssh) while the updater is running and you don't want the system to be rebooted under you in case an update gets pulled in, you should do a

```
    systemd-inhibit --what=shutdown $SHELL
```

Once you're done, you can simply exit the shell to let any potential reboot/updates to proceed.

To start using/testing ostree-based updates, you need to
1) enable ostree
2) expose the generated ostree archive-z2 repository over HTTP/HTTPS
3) configure the ostree remote to point to your exposed repository

I usually do this in the following way, although there are plenty of other ways, too (for instance, directly setting the ostree remote to your development/build machines IP address or host name in which case you don't need to fiddle with /etc/hosts on your test box).

- enable ostree support in local.conf (and enable password-less root logins):
```
REFKIT_IMAGE_EXTRA_FEATURES += "ostree debug-tweaks"
```
- override the default remote in local.conf, to use HTTP instead of HTTPS:
```
OSTREE_REMOTE = "http://update.example.org/ostree/"
```

- expose the ostree repo generated during the build using apache by adding this:
```
Alias "/ostree/" "<path-to-your-cloned-refkit-repo>/build/tmp-glibc/deploy/ostree-repo/"

<Directory <path-to-your-cloned-refkit-repo>/build/tmp-glibc/deploy/ostree-repo>
    Options Indexes FollowSymLinks
    Require all granted
</Directory>
```

- and finally on my test box running a now ostree-enabled image, I alias the remote to my host in /etc/hosts:
```
...
<build-machine IP address>    update.example.org              update
...
```

Then I just make sure my firewall lets HTTP traffic in, generate new builds and let my test box automatically pull the new builds in and reboot. If I need to, I use the systemd-inhibit trick above to prevent the box from rebooting when an update has been pulled in.
